tpu-client-next/src/node_address_service/leader_tpu_cache_service.rs

```diff
@@ -0,0 +1,516 @@
+/// for corresponding slot leaders.
+fn leader_sockets(
+    first_slot: Slot,
+    lookahead_leaders: u8,
+    slot_leaders: &SlotLeaders,
+    leader_tpu_map: &LeaderTpuMap,
+) -> Vec<SocketAddr> {
+    let fanout_slots = (lookahead_leaders as u64).saturating_mul(NUM_CONSECUTIVE_LEADER_SLOTS);
+    let mut leader_sockets = Vec::with_capacity(lookahead_leaders as usize);
+    // `slot_leaders.first_slot` might have been advanced since caller last read
+    // it. Take the greater of the two values to ensure we are reading from the
+    // latest leader schedule.
+    let current_slot = std::cmp::max(first_slot, slot_leaders.first_slot);
+    for leader_slot in
+        (current_slot..current_slot + fanout_slots).step_by(NUM_CONSECUTIVE_LEADER_SLOTS as usize)
+    {
+        if let Some(leader) = slot_leaders.slot_leader(leader_slot) {
+            if let Some(tpu_socket) = leader_tpu_map.get(leader) {
+                leader_sockets.push(*tpu_socket);
+                debug!("Pushed leader {leader} TPU socket: {tpu_socket}");
+            } else {
+                // The leader is probably delinquent
+                debug!("TPU not available for leader {leader}");
+            }
+        } else {
+            // Overran the local leader schedule cache
+            warn!(
+                "Leader not known for slot {}; cache holds slots [{},{}]",
+                leader_slot,
+                slot_leaders.first_slot,
+                slot_leaders.last_slot()
+            );
+        }
+    }
+
+    leader_sockets
+}
+
+async fn initialize_state(
+    rpc_client: &impl ClusterInfoProvider,
+    slot_receiver: SlotReceiver,
+    max_attempts: usize,
+) -> Result<(LeaderTpuMap, EpochInfo, SlotLeaders), Error> {
+    const ATTEMPTS_SLEEP_DURATION: Duration = Duration::from_millis(1000);
+    let mut leader_tpu_map = None;
+    let mut epoch_info = None;
+    let mut slot_leaders = None;
+    let mut num_attempts: usize = 0;
+    while num_attempts < max_attempts {
+        if leader_tpu_map.is_none() {
+            leader_tpu_map = LeaderTpuMap::new(rpc_client).await.ok();
+        }
+        if epoch_info.is_none() {
+            epoch_info = EpochInfo::new(rpc_client, slot_receiver.slot()).await.ok();
+        }
+
+        if let Some(epoch_info) = &epoch_info {
+            if slot_leaders.is_none() {
+                slot_leaders =
+                    SlotLeaders::new(rpc_client, slot_receiver.slot(), epoch_info.slots_in_epoch)
+                        .await
+                        .ok();
+            }
+        }
+        if leader_tpu_map.is_some() && epoch_info.is_some() && slot_leaders.is_some() {
+            return Ok((
+                leader_tpu_map.take().unwrap(),
+                epoch_info.take().unwrap(),
+                slot_leaders.take().unwrap(),
+            ));
+        }
+
+        num_attempts += 1;
+        sleep(ATTEMPTS_SLEEP_DURATION).await;
```

Comment:
*minor*

I would probably make this loop faster.
Considering we are looking at network operations, I do not think it makes sense to wait more than 100ms from the beginning of the loop iteration.

I suggest recording time at the loop start and then waiting only as long, as it would be necessary for the whole loop body to be 100ms.
Which would allow immediate retry if a single loop iteration took longer than that.